### PR TITLE
helm: initial support for profiles

### PR DESCRIPTION
**Pull request by howardjohn**
Part of https://github.com/istio/istio/issues/47838

in that you can now add `--set profile=ambient` or `--set profile=openshift` with helm.

How it works is we setup the "profile" as a file included in the helm chart. We then want to merge this on top of the `.Values` that gets passed into templates.

Helm executes templates in reverse order by filename, as we make a `zzz_profile.yaml` to ensure it runs first. In that file, we read the appropriate profile and merge it on `.Values`. Then, all other templates access values like normal.

We will want these profiles in each chart; I currently have them in istio-discovery only. We can have a makefile target that syncs them across all charts (we do this for other things as well).

For most charts, all values are prefixed (`.pilot.xyz`). However, some are not (especially under `.global`). In theory, we might want to pass a different value for the same `.foo` to different charts. There is no case _today_ when we need to do this, but _if_ we do, we will just make it so the profile is different for the different charts (and make the makefile accommodate this); not a blocker. Because we use a single file for all charts, this does mean we set keys under `.pilot` for ztunnel (for example); this is a bit odd but has no real impact.

A major difference with istioctl profiles (possibly the only difference?) is that istioctl also picks which components are enabled. Helm users do that themselves via multiple `helm install` commands. We can dedupe the code here in istioctl to just set this and `values.profile`. Example of ambient IOP profile:

```yaml
spec:
  components:
    cni:
      enabled: true
    ztunnel:
      enabled: true
    ingressGateways:
    - name: istio-ingressgateway
      enabled: false
values:
  profile: ambient
```

Today, we just allow users to set 1 profile, matching Istioctl. A follow up to this will probably also allow setting a `compatibility-version` field per https://docs.google.com/document/d/12c-0SBIfLJ4UceI9_UkyLNhPxjrT-Y0I1FsdhAhtIl8/edit#heading=h.xw1gqgyqs5b

This depends on https://github.com/istio/release-builder/pull/1717. I will merge that once this is approved.

### CODE REVIEW

**Comment by keithmattix** [comment]
The direction looks elegant! +1 to compatibility-version being added in the future

**Comment by kfaseela** [comment]
Thanks @howardjohn for this work!

**Comment by jacob-delgado** [comment]
This looks like a great direction for helm. 👍🏽

**Comment by therealmitchconnors** [comment]
Not an env maintainer, but I am really excited to fix this for users.  It came up several times in our recent survey!  👍

**Comment by howardjohn** [comment]
blocked by https://github.com/istio/istio/pull/48332 (unless we want to decouple them)

File: manifests/charts/base/README.md
Comment (hanxiaop) [inline manifests/charts/base/README.md:27]:
```suggestion
For example, the `demo` profile offers a preset configuration to try out Istio in a test environment, with additional features enabled and lowered resource requirements.
```

Comment (hanxiaop) [reply manifests/charts/base/README.md:27]: I think other charts' README files have the same issue.

File: manifests/charts/base/templates/zzz_profile.yaml
Comment (hanxiaop) [inline manifests/charts/base/templates/zzz_profile.yaml:2]: Interesting approach! Should we briefly include the merge logic in the README as well?

**Comment by dgn** [comment]
Great idea. I think this will allow for some simplification of the istioctl code base. Regarding the component enabling... Can't that be done through helm as well? Why would we need to do it again in the components section?

**Comment by howardjohn** [comment]
> Great idea. I think this will allow for some simplification of the istioctl code base. Regarding the component enabling... Can't that be done through helm as well? Why would we need to do it again in the components section?

My understanding: Components and values are quite different. In helm, enabling a component means running a completely separate `helm install <component> -f values`. Istioctl does multiple of these in one go, hence the components.

Technically I think you can make a composite/monolith helm chart that has sub charts enabled via values to give similar behavior to istioctl, though I am not sure we want to do that.

**Comment by rcernich** [comment]
> Technically I think you can make a composite/monolith helm chart that has sub charts enabled via values to give similar behavior to istioctl, though I am not sure we want to do that.

Just some history...

I think that's how the charts were initially structured (back in the galley/mixer days).  The challenge with that approach was that you ended up with a lot of duplicate fields, because the values for each component were isolated (e.g. `Values.component.foo`) and could not be referenced across components., hence `Values.global`.  That said, I think some of the challenges that came out of the move away from a monolithic set of charts was that it became more complicated to get a "complete" Istio install, and a decision was made not to support any kind of shell script, hence the istio command line operator was born.  I'm not sure how much of a problem this is now that the control plane is much simpler and third-party extensions have basically been removed from the main charts.  FWIW

**Comment by howardjohn** [comment]
> > Technically I think you can make a composite/monolith helm chart that has sub charts enabled via values to give similar behavior to istioctl, though I am not sure we want to do that.
> 
> Just some history...
> 
> I think that's how the charts were initially structured (back in the galley/mixer days). The challenge with that approach was that you ended up with a lot of duplicate fields, because the values for each component were isolated (e.g. `Values.component.foo`) and could not be referenced across components., hence `Values.global`. That said, I think some of the challenges that came out of the move away from a monolithic set of charts was that it became more complicated to get a "complete" Istio install, and a decision was made not to support any kind of shell script, hence the istio command line operator was born. I'm not sure how much of a problem this is now that the control plane is much simpler and third-party extensions have basically been removed from the main charts. FWIW

:+1: :+1: I will also say tooling around Helm has improved for this as well. Things like helmfile, argocd, etc. Even terraform

**Review by hanxiaop — APPROVED** [review]

**Review by linsun — COMMENTED** [review]
LGTM - thank you for this great simplification for helm users @howardjohn 

One q - does this mean when we need to change any of the profile (for example ambient), we need to change in multiple places? I saw profile-ambient.yaml in a few folders. Could be error prune for us to modify any of the profiles.

**Comment by howardjohn** [comment]
> LGTM - thank you for this great simplification for helm users @howardjohn
> 
> One q - does this mean when we need to change any of the profile (for example ambient), we need to change in multiple places? I saw profile-ambient.yaml in a few folders. Could be error prune for us to modify any of the profiles.

No, you only change the top level one, automation copies it to other places. We do this for other parts of the charts already

File: manifests/charts/base/files/profile-demo.yaml
Comment (hzxuzhonghu) [inline manifests/charts/base/files/profile-demo.yaml:19]: These should be installed as needed

File: manifests/charts/base/values.yaml
Comment (hzxuzhonghu) [inline manifests/charts/base/values.yaml:1]: nit: why add a new level `defaults`

File: manifests/charts/base/files/profile-demo.yaml
Comment (hanxiaop) [reply manifests/charts/base/files/profile-demo.yaml:19]: This is for otel demos: https://github.com/istio/istio/tree/master/samples/open-telemetry

Comment (howardjohn) [reply manifests/charts/base/files/profile-demo.yaml:19]: :+1: its a demo profile to install a bunch of stuff for you so you can go through all the istio.io docs. Its a direct copy from the istioctl profile, nothing new -- just adding support for Helm.

File: manifests/charts/base/values.yaml
Comment (howardjohn) [reply manifests/charts/base/values.yaml:1]: Its described in the README in the PR and zzz_profile.yaml

```
We have three sets of values, in order of precedence (last wins):
1. The builtin values.yaml defaults
2. The profile the user selects
3. Users input (-f or --set)

Unfortunately, Helm provides us (1) and (3) together (as .Values), making it hard to insert (2).

However, we can workaround this by placing all of (1) under a specific key (.Values.defaults).
We can then merge the profile onto the defaults, then the user settings onto that.
Finally, we can set all of that under .Values so the chart behaves without awareness.
```

File: manifests/charts/base/files/profile-demo.yaml
Comment (zirain) [reply manifests/charts/base/files/profile-demo.yaml:19]: this's very useful to simply test setup for istio.io.

**Review by keithmattix — APPROVED** [review]
LGTM

**Review by zirain — APPROVED** [review]

File: manifests/charts/base/files/profile-preview.yaml
Comment (hzxuzhonghu) [inline manifests/charts/base/files/profile-preview.yaml:11]: remove it

File: manifests/charts/default/files/profile-preview.yaml
Comment (hzxuzhonghu) [inline manifests/charts/default/files/profile-preview.yaml:11]: re,move

**Review by ericvn — APPROVED** [review]

**Comment by stevehipwell** [comment]
@howardjohn how come this PR included changes to the operator Helm chart that don't actually do anything?

**Comment by howardjohn** [comment]
The intent was to change all charts for consistency so that if we add or
change profiles to touch other charts we do not need to change the core of
the chart.

Is it causing issues?

**Comment by stevehipwell** [comment]
@howardjohn as the operator chart doesn't currently do anything with profiles it's unnecessary and makes the code harder to understand, especially if you don't know that the profiles are irrelevant.

If this logic was being added to the operator chart (in combination with creating the CR) none of the existing values need merging as they're not relevant to the profile merging logic; you'd be looking at manipulating the `installation.spec` value given the hypothetical additional values below.

```yaml
installation:
  # -- If `true`, create an `IstioOperator` CR.
  enabled: false
  
  # -- `IstioOperator` CR spec.
  spec:
```

Out of interest why do you need an intermediate key of `default` for the merge logic? It looks like a SemVer major change, adds complexity and is very likely to cause issues due to the change and it not being idiomatic. Wouldn't interacting directly with `.Values` and either filtering out (deny list) or filtering in (allow list) work the same but not require a user facing API change?

**Comment by howardjohn** [comment]
Can you help me understand how it breaks compatibility? users still interact with the chart setting the exact same values as before. All the (complex) merging stuff is an implementation detail.

I wrote a bit more about this here https://blog.howardjohn.info/posts/advanced-helm/#profiles

**Comment by stevehipwell** [comment]
@howardjohn sorry I missed the actual pattern here; but I think that just highlights my point about it being non-idiomatic and confusing. For the record I maintain a number of well used Helm charts and utilise the style of value modification you've implemented here, so if I missed the nuance here I wont be alone.

So let me update my position here as it's not a breaking change as I incorrectly suggested above in the ~~struck comments~~.

1. The operator chart doesn't need a `defaults` as none of the values are present in a profile to need a 3 way merge.
2. The `defaults` pattern is confusing as the idiomatic pattern for updating a Helm chart is to take the values and modify them with new values, the only values are now under `defaults` so those will be the ones modified.

I do like the pattern here, now I've groked it, but wouldn't it make sense for `defaults` to be in addition to the standard chart values (see below) and below the actual values?

```yaml
value1: hello
value2: world

# Do not modify the defaults
defaults:
  value2: world
```

This could be manually managed but this looks like a task that could be automated (Helm plugin even); if you grab the top level keys from each chart profile you could figure out which values keys needed copying under `defaults` as part of the chart release process.

**Comment by stevehipwell** [comment]
As it stands (correct me if I'm wrong) consumers checking the GitHub source or published chart (via `helm pull` or `helm show values`) would not see the values they should be overriding. The values aren't documented in the README (either manually or via `helm-docs`). And the defaults in the values don't warn the user not to modify them.

The last mile mutation pattern would make the 3 way merge work by replicating the values needing a 3 way merge into the `defaults` value while leaving the rest of the chart idiomatic and able to be queried by the user. I'd also suggest documenting the non defaults values with `helm-docs` and annotating `defaults` to not be included (implementation detail).

**Comment by howardjohn** [comment]
Ah I see, that does seem like it would help. Though last mile may make it challenging for development purposes/test which do use the GitHub source directly.

I think @bleggett was looking into documentation for the helm charts possibly (or at least wanted them to exist.. 0

**Pull request merged**
